### ISSUE TITLE
Missing documentation for mask format for segmentation IOU type in mAP detection metric

### ISSUE THREAD

**Issue by JeroenDelcour** [Lightning-AI/metrics#1061]
The [documentation](https://torchmetrics.readthedocs.io/en/v0.9.0/detection/mean_average_precision.html) seems to be missing the required format for masks when using the "segm" option for `iou_type`.

I'm assuming there has to be a dict entry with key "masks", but what should the shape and dtype of the masks be? I tried a few different options, but always got cryptic errors from pycocotools. The [example script](https://github.com/PyTorchLightning/metrics/blob/567fa3405d5a60ba0a6052e57afe5a5d34c376a7/tm_examples/detection_map.py) doesn't mention masks at all. I tried to figure it out from the [unit tests](https://github.com/PyTorchLightning/metrics/blob/567fa3405d5a60ba0a6052e57afe5a5d34c376a7/tests/detection/test_map.py), but I couldn't really follow what was going on.

I'd be happy to create a PR and improve the documentation myself if someone can help me figure out the required format.

Related question: I was very much looking forward to the torchmetrics implementation of mAP with masks because I thought it meant not having to deal with the poorly documented pycocotools. I've been following #822 and saw pycocotools is still required. Besides improving documentation, are there plans for removing this dependency entirely?

**Comment by SkafteNicki** [comment]
cc: @gianscarpe can you comment what the format should be then @JeroenDelcour can do the actual PR? :]

**Comment by gianscarpe** [comment]
Hi! let me clarify this.
Predictions are a list of dictionaries. Each dictionary has 3 keys, "masks", "labels", and "scores". Labels and scores are the same as in bounding-boxes IOU. "masks" is a tensor of shape N_predxHxW, as the example below
```
{'masks': tensor([[[False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False],
         ...,
         [False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False]],

        [[False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False],
         ...,
         [False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False],
         [False, False, False,  ..., False, False, False]]]), 'scores': tensor([0.3180, 0.7260]), 'labels': tensor([3, 2], dtype=torch.int32)}
```
In unit-test was easier to use the RLE version and decode it when giving it as input to `update`. Thanks for pointing out!

**Comment by gianscarpe** [comment]
Regarding the dependency, unfortunately, mask-IOU is time and computation-consuming, and pycocotools have a CPP implementation. What I suggest I implementing a RLE toolbox that also works with CUDA and integrate it in torchmetrics, removing pycocotools depency

**Comment by gianscarpe** [comment]
@23pointsNorth thanks for trying this out! It seems from  your reported issue that you're using IOU type bboxes. Can you check at initialization what IOU type are you passing as argument? should be "segm" for using segmentation IOU
```
File "/usr/local/lib/python3.8/site-packages/torchvision/ops/boxes.py", line 241, in box_area
  boxes = _upcast(boxes)
```

**Comment by 23pointsNorth** [comment]
Hi @gianscarpe , it was the case that if there are 0 predictions made by the model, another call was made that had some leftover area calculation that wasn't converted to use the `self.iou_type`. See here for the fix #1080. 

Thanks for doing #822 in the first place!

**Comment by faber6911** [comment]
Hi @gianscarpe,
one thing is not clear to me about the input format.

The masks must necessarily contain only one element for each mask?

e.g.
```python
import torch

masks = torch.tensor(
    [
        [
             [0, 0, 0, 0, 0],
             [0, 0, 1, 1, 0],
             [0, 0, 1, 1, 0],
             [0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0],
        ],
    ],
)
```
or it could be also something like
```python
import torch

masks = torch.tensor(
    [
        [
             [1, 1, 0, 0, 0],
             [1, 1, 0, 0, 0],
             [0, 0, 0, 0, 0],
             [1, 1, 0, 1, 1],
             [1, 1, 0, 0, 0],
        ],
    ],
)
```

**Comment by gianscarpe** [comment]
Hi @faber6911, thanks for reaching out. I believe you're referring to multi-predictions, but correct me if I'm wrong. In that case, you should have a mask of N channels (where N number of predictions), similarly to what it's already done with bounding-boxes. So that:
```
masks = torch.tensor(
    [
        [
             [0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0],
             [0, 0, 0, 1, 1],
             [0, 0, 0, 0, 0],
        ], # mask for pred. 1
        [
             [1, 1, 0, 0, 0],
             [1, 1, 0, 0, 0],
             [0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0],
        ], # mask for pred. 2
    ],
)

labels = torch.tensor([1, 2]) # labels for preds 
```

**Comment by faber6911** [comment]
@gianscarpe Yep, thank you.
With bboxes it was easier for me to figure out how to feed the metric but with segmentation maps, it was difficult for me to have the same intuition.

**Comment by JeroenDelcour** [comment]
Hi, sorry for not responding sooner. Thank you @gianscarpe for explaining. I was hoping to have the time to help write the documentation, but unfortunately priorities at my work have changed. I'd still like to help, but there's not telling when I might have the time. Apologies!

**Comment by JeroenDelcour** [comment]
I found some time to do this after all :)
